pagopa/io-app
# chore: [IOPID-192] CIE 409 errors

## Short description
This PR adds to the CIE login a new way to handle redirects and errors via native library.
It also adds http error handling to login WebViews.

## List of changes proposed in this pull request
- [package.json](https://github.com/pagopa/io-app/pull/4619/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519): added `io-react-native-login-utils` library
-  [ts/components/cie/CieRequestAuthenticationOverlay.tsx](https://github.com/pagopa/io-app/pull/4619/files#diff-eb0dafddc14b9bf7689c3ee821d1b5c30202a8077cb4e24288e90195318816f3): 
     - Added new states to the view
     - Adapted handleOnError function to idp login error style
     - JS injecting is now properly done only for iOS devices
     - Added function `regenerateKeyGetRedirectsAndVerifySaml` that use the native library to handle redirects

- [ts/features/lollipop/index.ts](https://github.com/pagopa/io-app/pull/4619/files#diff-a773fe09de09754a0f2fee0057f3e318b00c8049f8cf69b09d23377752208e3c): added a function that regenerates lollipop key
- [[ts/features/lollipop/utils/__tests__/retryAndVerifyMismatch.test.ts](https://github.com/pagopa/io-app/pull/4619/files#diff-bb861bbccfb56860f135b447d8c9bd3e37efade177c2bb0415eed7fa83d5e737),[ts/features/lollipop/utils/__tests__/retryAndVerifyMissingSamlRequest.test.ts](https://github.com/pagopa/io-app/pull/4619/files#diff-44907fc1c12ff2d67ec321ae2e6d086c1f47f734764a16255b1ff8ceb93b6ce9),[ts/features/lollipop/utils/__tests__/retryAndVerifyMissingThumbprint.test.ts](https://github.com/pagopa/io-app/pull/4619/files#diff-6bfad00007ef648f2949fb0a533dd4797c761f331e8a0cd31ccb14dc0923bcc5),[ts/features/lollipop/utils/__tests__/retryAndVerifyRedirectNativeError.test.ts](https://github.com/pagopa/io-app/pull/4619/files#diff-38a556108890cad3c4d7a4e811cce916ddc46594c4edc0e73b6b6ff585635e7b),[ts/features/lollipop/utils/__tests__/retryAndVerifySuccess.test.ts](https://github.com/pagopa/io-app/pull/4619/files#diff-7eb65dbbe0c949778f920804fe110695b3983cdcac71d5600e870100b231630b),[ts/features/lollipop/utils/__tests__/saml.test.ts](https://github.com/pagopa/io-app/pull/4619/files#diff-7301c8b01c6d8ff8e124e10e550a623fffedf96fa7251055b5c8092095c6af0d)]: tests for new functions
- [ts/features/lollipop/utils/login.ts](https://github.com/pagopa/io-app/pull/4619/files#diff-f54ddab5813a42b9135c11cbd42b931854b83272a0763ec2fd2b759f8ede6f09):
    - Added new try/catch element to `lollipopSamlVerify`
    - Added `regenerateKeyGetRedirectsAndVerifySaml` implementation. 
- [ts/screens/authentication/IdpLoginScreen.tsx](https://github.com/pagopa/io-app/pull/4619/files#diff-42db30685d69d36c270f8b986b1e400960ab1c85ec3e9a79f8eecf2007ee8c89): added http error handling
- [ts/screens/authentication/cie/CieConsentDataUsageScreen.tsx](https://github.com/pagopa/io-app/pull/4619/files#diff-89c251a9a9539e3470c6001c13917f0881272bfa692f61bdc4a6f191b0435fa3): added originWhiteList to WebView
- [ts/utils/analytics.ts](https://github.com/pagopa/io-app/pull/4619/files#diff-a425aab9658df7526ca2afd44dfd12ab19227f77c49e80436ef35388f2266627): unified management of login errors
- [[ts/features/lollipop/hooks/useLollipopLoginSource.tsx](https://github.com/pagopa/io-app/pull/4619/files#diff-a66c588312e074ecef0c31f39668b6097c22fd8745e765b489150f3c9d165c5b), [ts/features/lollipop/index.ts](https://github.com/pagopa/io-app/pull/4619/files#diff-a773fe09de09754a0f2fee0057f3e318b00c8049f8cf69b09d23377752208e3c)]: refactored to unify the handling of functions needed for the native login process as well

## How to test

Run tests.
Try to login with cie and idp.
With proxyman (or whatever), try to get the login calls wrong and verify that the errors are properly handled.
